src/jsMain/kotlin/process/phonemes/PhonemesMapping.kt

```diff
@@ -48,20 +132,22 @@ fun Track.replacePhonemes(request: PhonemesMappingRequest?) = copy(
 
 fun Note.replacePhonemes(request: PhonemesMappingRequest?): Note {
     if (request == null) return copy(phoneme = null)
-    var output = phoneme?.split(" ") ?: return this
+    var rawInput = phoneme?.split(" ") ?: return this
+    var input = rawInput.map { it to true }.toMutableList() // boolean represents the phoneme is not converted yet
+    var output = listOf<Pair<String, Int>>() // int represents index from input phoneme list
     for ((key, value) in request.map) {
         val keySplit = key.split(" ")
-        for (i in 0..(output.size - keySplit.size)) {
-            val subList = output.subList(i, i + keySplit.size)
-            if (subList == keySplit) {
-                output = output.subList(0, i) + value.split(" ") +
-                    output.subList(i + keySplit.size, output.size)
-                // once replaced, the count of output is changed, so we could not proceed the replacement.
-                // this means that multiple occurrences of the same phoneme set in one note get replaced only once.
-                break
+        for (i in 0..(input.size - keySplit.size)) {
+            val subList = input.subList(i, i + keySplit.size)
+            if (subList.map { it.first } == keySplit && subList.map { it.second }.all { it }) {
+                output += Pair(value, i)
+                for (j in 0..subList.size - 1) {
+                    subList[j] = Pair(subList[j].first, false)
```

Comment:
`subList` is a local `val` inside `for (i in 0..(input.size - keySplit.size))` scope, here, I think we every phoneme should only be used once in the whole replacement process? So we would need to update the `input` list as well?

Reply:
`subList` is a shallow copy of elements from `input` so the updates are still reflected in the main `input` list

Reply:
Ah, yes, it's just a view of the original list, so when you do `subList[j]=..` it changes the `input` as well 👍 
<img width="380" alt="image" src="https://github.com/sdercolin/utaformatix3/assets/7665216/d10d4fad-30a9-4984-8f37-286a5bd9c839">

(I think we probably don't call it a "shallow copy" actually, if you want to do a shallow copy, to need to call `toMutableList` again.)